### ISSUE TITLE
Allow server application to decide what to do when no client certificate is presented

### ISSUE THREAD

**Issue by anrossi** [microsoft/msquic#1953]
Schannel and OpenSSL client certificate support will fail the handshake if the server requires a client certificate be presented during the handshake, and no client certificate is presented.
In some scenarios, the server would prefer to know this and decide what to do instead of the TLS layer killing the handshake.

**Comment by anrossi** [comment]
A workaround exists, however: If `QUIC_CREDENTIAL_FLAG_REQUIRE_CLIENT_AUTHENTICATION` is not set on the server configuration, the server application can set `QUIC_CREDENTIAL_FLAG_INDICATE_CERTIFICATE_RECEIVED` and verify that it receives a client certificate during the handshake, and if not, it can decide what to do from there.

**Comment by wfurt** [comment]
Related to https://github.com/dotnet/runtime/issues/57308
The server will _not_ ask for certificate without `QUIC_CREDENTIAL_FLAG_REQUIRE_CLIENT_AUTHENTICATION` right @anrossi? With that I think client will never send it.

**Comment by anrossi** [comment]
@wfurt My understanding is not complete here; won't the client send a client certificate if it is configured with one, even if the server doesn't ask for it? Or does client wait to send the client certificate until requested by the server?

**Comment by wfurt** [comment]
That depends - and it should be visible in packet capture. My experience is that it won't until it gets 
[Certificate Request](https://datatracker.ietf.org/doc/html/rfc8446#section-4.3.2).

**Comment by nibanks** [comment]
Let's add a new flag that says "Indicate NULL cert if not present".

**Comment by wfurt** [comment]
Here are some more references: 
Tls 1.2
https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.6
```
      This is the first message the client can send after receiving a
      ServerHelloDone message.  This message is only sent if the server
      requests a certificate.  If no suitable certificate is available,
      the client MUST send a certificate message containing no
      certificates.  That is, the certificate_list structure has a
      length of zero.  If the client does not send any certificates, the
      server MAY at its discretion either continue the handshake without
      client authentication, or respond with a fatal handshake_failure
      alert.
```
This is the general behavior we looking at. I could not find quite same text in Tls 1.3 e.g. it is not clear to me if client _can_ send certificate without server asking for it. 

https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2.4
```
   If the server supplies an empty Certificate message, the client MUST
   abort the handshake with a "decode_error" alert.

   If the client does not send any certificates (i.e., it sends an empty
   Certificate message), the server MAY at its discretion either
   continue the handshake without client authentication or abort the
   handshake with a "certificate_required" alert.  Also, if some aspect
   of the certificate chain was unacceptable (e.g., it was not signed by
   a known, trusted CA), the server MAY at its discretion either
   continue the handshake (considering the client unauthenticated) or
   abort the handshake.
```

The default behavior is to fail if cert was requested but not provided. 
But with current SslStream and Kestrel we allow to pass that MAY to user callback. 

New flag would be ok. We also set `QUIC_CREDENTIAL_FLAG_NO_CERTIFICATE_VALIDATION` and we _may_ use it as indication that msquic should not worry or error out if the cert was not provided.

**Comment by anrossi** [comment]
Fixed in #2702 and #2738
